pkg/engine/mutate/patchesUtils.go

```diff
@@ -34,6 +33,123 @@ func generatePatches(src, dst []byte) ([][]byte, error) {
 	return patchesBytes, err
 }
 
+// sortRemovalPatches sort the removal patches(with same path) by the key of index in descending order
+//
+// [{"op":"remove","path":"/a/b/0"},{"op":"remove","path":"/a/b/1"}]
+// will be reordered to
+// [{"op":"remove","path":"/a/b/1"},{"op":"remove","path":"/a/b/0"}]
+func sortRemovalPatches(originalPatches []jsonpatch.JsonPatchOperation) []jsonpatch.JsonPatchOperation {
+	patches := filterInvalidPatches(originalPatches)
+
+	result := make([]jsonpatch.JsonPatchOperation, len(patches))
+	index := getIndexToBeReversed(patches)
+
+	if len(index) == 0 {
+		return patches
+	}
+
+	start := 0
+	for _, idx := range index {
+		end := idx[0]
+		copy(result[start:end], patches[:end])
+		reversedPatches := reverse(patches, idx)
+		copy(result[end:], reversedPatches)
+		start = idx[1] + 1
+	}
+	copy(result[start:], patches[start:])
+	return result
+}
+
+func getIndexToBeReversed(patches []jsonpatch.JsonPatchOperation) [][]int {
+	removePaths := make([]string, len(patches))
+
+	for i, patch := range patches {
+		if patch.Operation == "remove" {
+			removePaths[i] = patch.Path
+		}
+	}
+	return getRemoveInterval(removePaths)
+
+}
+
+func getRemoveInterval(removePaths []string) [][]int {
+	// get paths end with '/number'
+	regex := regexp.MustCompile(`\/\d+$`)
+	for i, path := range removePaths {
+		if !regex.Match([]byte(path)) {
+			removePaths[i] = ""
+		}
+	}
+
+	res := [][]int{}
+	for i := 0; i < len(removePaths); {
+		if removePaths[i] != "" {
+			baseDir := filepath.Dir(removePaths[i])
+			j := i + 1
+			for ; j < len(removePaths); j++ {
+				curDir := filepath.Dir(removePaths[j])
+				if baseDir != curDir {
+					break
+				}
+			}
+			if i != j-1 {
+				res = append(res, []int{i, j - 1})
+			}
+			i = j
+		} else {
+			i++
+		}
+	}
+
+	return res
+}
+
+func reverse(patches []jsonpatch.JsonPatchOperation, interval []int) []jsonpatch.JsonPatchOperation {
+	res := make([]jsonpatch.JsonPatchOperation, interval[1]-interval[0]+1)
+	j := 0
+	for i := interval[1]; i >= interval[0]; i-- {
+		res[j] = patches[i]
+		j++
+	}
+	return res
+}
+
+// filterInvalidPatches filters out patch with the following path:
+// - not */metadata/name, */metadata/namespace, */metadata/labels, */metadata/annotations
+// - /status
+func filterInvalidPatches(patches []jsonpatch.JsonPatchOperation) []jsonpatch.JsonPatchOperation {
```

Comment:
Is it a safe assumption to always remove these patches? For the ones where we had `nil` values, I was thinking it would be best to check the source data and see if the value was `nil` and remove only then...